wip/wip.p8

```diff
@@ -79,105 +85,119 @@ function draw(ent)
 end
 
 function _update()
- for i, ent in ipairs(ents) do
-  if(ent.update) ent:update()
- end
+  for i, ent in ipairs(ents) do
+    if(ent.update) ent:update()
+  end
 end
 
 
 function _draw()
- cls(0)
- map(0, 0, 0, 0, 128,64)
- for i, ent in ipairs(ents) do
-  draw(ent)
- end
+  cls(0)
+  map(0, 0, 0, 0, 128,64)
+  for i, ent in ipairs(ents) do
+    draw(ent)
+  end
 end
 
--->8
 function norm(val)
- return val/abs(val)
+  if(val==0) return 0
+  return val/abs(val)
 end
 
 function clamp(val,val_max)
- if abs(val)<val_max then
-	 return val
-	else
-	 return val_max*norm(val)
- end
+  if abs(val)<val_max then
+    return val
+  else
+    return val_max*norm(val)
+  end
 end
 
 function accel(ent,amt,direc)
- if     direc=="x" then
-  ent.v_x=clamp(ent.v_x+amt,ent.v_max)
- elseif direc=="y" then
-  ent.v_y=ent.v_y+amt
+  if     direc=="x" then
+    ent.v_x=clamp(ent.v_x+amt,ent.v_max)
+  elseif direc=="y" then
+    ent.v_y=ent.v_y+amt
 	end
 end
 
 function apply_fric_sub(v)
- local normal=norm(v)
- local amp=abs(v)
- local new=amp-world.friction
- if new<0 then new=0 end
- return new*normal
+  local normal=norm(v)
+  local amp=abs(v)
+  local new=amp-world.friction
+  if new<0 then new=0 end
+  return new*normal
 end
 
 function apply_fric(ent)
-	ent.v_x=apply_fric_sub(ent.v_x)
- ent.v_y=apply_fric_sub(ent.v_y)
+  ent.v_x=apply_fric_sub(ent.v_x)
+  ent.v_y=apply_fric_sub(ent.v_y)
 end
 
 function apply_g(ent)
- ent.v_y=ent.v_y+world.g
+  ent.v_y=ent.v_y+world.g
 end
 
-function constrain_to_screen(ent)
- local x=ent.x
- local y=ent.y
- local bbox=ent.bbox
- if(x<0) x=0
- if(x+bbox.w>127) x=127-bbox.w
- if(y<0) y=0
- if(y+bbox.h>127) y=127-bbox.h 
- ent.x=x
- ent.y=y
+function map_collision(x,y,points)
+  local collision=false
+  for i,point in ipairs(points) do
+    local p_x,p_y=unpack(point)
+    local tile=mget((x+p_x)/8,(y+p_y)/8)
+    if(fget(tile,0)) then
+      collision=true
+    end
+  end
+  return collision
 end
 
-function map_collision(x,y,w,h)
-	local top_l=mget(flr(x/8),flr(y/8))
- local bot_l=mget(flr(x/8),flr((y+h)/8))
- local top_r=mget(flr((x+w)/8),flr(y/8))
- local bot_r=mget(flr((x+w)/8),flr((y+h)/8))
-	 
-	return fget(top_l)==1 
-	    or fget(bot_l)==1 
-	    or fget(top_r)==1 
-	    or fget(bot_r)==1
+function chk_collision(ent, collision_fn)
+  -- 1=top,2=bottom,3=left,4=right
+  for dir=1,4 do
+    if(dir==1 and ent.v_y>0) then goto continue end
+    if(dir==2 and ent.v_y<0) then goto continue end
+    if(dir==3 and ent.v_x>0) then goto continue end
+    if(dir==4 and ent.v_x<0) then goto continue end
+
+    local vec_length=sqrt(ent.v_x*ent.v_x + ent.v_y*ent.v_y)
+    local segment=0
+    local proj_x=0
+    local proj_y=0
+    while(not collision_fn(ent.x+proj_x,
+                            ent.y+proj_y,
+                            ent.bbox[dir])
+          and segment<vec_length) do
+      proj_x+=ent.v_x/vec_length
+      proj_y+=ent.v_y/vec_length
+      segment+=1
+    end
+    if(segment<vec_length) then
```

Comment:
For velocities less than abs(1) where the entity is next to obstruction the above `while` exits with a velocity that will take it into the obstruction, but the segment value is such that no correction will occur. Is this a problem?